feat(codedeploy): loadbalancer support for imported Target Groups

eastNine opened the pull request:
This PR fixes that imported alb and nlb target group be able to configure to loadBalancer property.

Fixes #9677.

example:

```TypeScript
import * as cdk from '@aws-cdk/core';
import * as codedeploy from '@aws-cdk/aws-codedeploy';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';

const deploymentGroup = new codedeploy.ServerDeploymentGroup(this, 'deploymentGroup', {
    ...

    // configurable imported application loadbalancer targetgroup
    loadBalancer: codedeploy.LoadBalancer.application(
        elbv2.ApplicationTargetGroup.fromTargetGroupAttributes(this, 'importedAlbTg', {
            targetGroupArn: 'arn:aws:elasticloadbalancing:ap-northeast-2:111111111111:targetgroup/myAlbTargetgroup/abcd12345678efgf'
        })
    ),

    // also network loadbalancer targetgroup
    loadBalancer: codedeploy.LoadBalancer.network(
        elbv2.NetworkTargetGroup.fromTargetGroupAttributes(this, 'importedNlbTg', {
            targetGroupArn: 'arn:aws:elasticloadbalancing:ap-northeast-2:111111111111:targetgroup/myNlbTargetgroup/wxyz09876543opqr'
        })
    ),
});
```

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 4562daa5206f2f3f047ba12c9b4fb01cc349ddfa
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ac3228ce-925f-43de-8d47-614d30372a08%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 8187bfb106f88415d6c0140e2a1b335af5ae7423
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0e1a8d3d-45a3-4581-b353-6803a55576f7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-codedeploy/test/server/deployment-group.test.ts:
```diff
@@ -215,9 +222,16 @@ describe('CodeDeploy Server Deployment Group', () => {
+                },
               ],
             },
           },
```
We're missing a unit test for using `LoadBalancer` with imported Target Groups here 🙂.

skinny85 commented on packages/@aws-cdk/aws-codedeploy/lib/server/load-balancer.ts:
```diff
@@ -41,10 +42,10 @@ export abstract class LoadBalancer {
     class AlbLoadBalancer extends LoadBalancer {
       public readonly generation = LoadBalancerGeneration.SECOND;
-      public readonly name = albTargetGroup.targetGroupName;
+      public readonly name = targetGruopNameFromArn(albTargetGroup.targetGroupArn);
```
I don't particularly like this solution. I actually think this is putting the logic of how to get the name of the Target Group from its ARN in the wrong place.

Let's do this a little bit differently:

1. Let's add a `targetGroupName` property to `ITargetGroup`.
2. In imported Target Groups, let's add the logic of parsing it from the ARN of the Target Group.
3. Let's keep using `public readonly name = albTargetGroup.targetGroupName;` here, but now, this will Just Work.

eastNine replied:
OK, I agree your proposal.👌
I will update soon with this approach.

skinny85 reviewed (changes requested):
Thanks for the contribution @eastNine! I think we should go a slightly different route than what you proposed here.

Let me know if you agree with my proposal!

Thanks,
Adam

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a8c8e885ecefa7fe234a0772370300ba196af218
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8e96d2a2-ecd4-4c18-bad3-2abfb9f33d66%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/util.ts:
```diff
@@ -90,3 +91,11 @@ export function mapTagMapToCxschema(tagMap: Record<string, string>): cxschema.Ta
+ * Return the targetGroupName from targetGroupArn
+ */
+export function targetGroupNameFromArn(targetGroupArn: string) {
+  const arnParts = Fn.split('/', targetGroupArn);
```
We have a function in the `core` Stack for this already: `Stack.splitArn()`. See how it's used [here](https://github.com/aws/aws-cdk/blob/203c42b792e46536eaf21fa48277985c3409b65c/packages/%40aws-cdk/aws-kinesisanalytics-flink/lib/application.ts#L230), for example.

skinny85 commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/imported.ts:
```diff
@@ -29,6 +33,7 @@ export abstract class ImportedTargetGroupBase extends CoreConstruct implements I
     super(scope, id);
 
     this.targetGroupArn = props.targetGroupArn;
+    this.targetGroupName = targetGroupNameFromArn(props.targetGroupArn) || cdk.Aws.NO_VALUE;
```
I don't think you need the `||` part (`targetGroupNameFromArn` never returns `undefined`).

skinny85 commented on packages/@aws-cdk/aws-codedeploy/test/server/deployment-group.test.ts:
```diff
@@ -412,4 +412,90 @@ describe('CodeDeploy Server Deployment Group', () => {
+    new codedeploy.ServerDeploymentGroup(stack, 'DeploymentGroup', {
+      loadBalancer: codedeploy.LoadBalancer.network(
+        lbv2.NetworkTargetGroup.fromTargetGroupAttributes(stack, 'importedNlbTg', {
+          targetGroupArn: targetGroup.targetGroupArn,
```
Can we write a test with a static ARN instead?

skinny85 commented on packages/@aws-cdk/aws-codedeploy/test/server/deployment-group.test.ts:
```diff
@@ -412,4 +412,90 @@ describe('CodeDeploy Server Deployment Group', () => {
+    new codedeploy.ServerDeploymentGroup(stack, 'DeploymentGroup', {
+      loadBalancer: codedeploy.LoadBalancer.application(
+        lbv2.ApplicationTargetGroup.fromTargetGroupAttributes(stack, 'importedAlbTg', {
+          targetGroupArn: targetGroup.targetGroupArn,
```
Can we write a test with a static ARN instead?

skinny85 commented on packages/@aws-cdk/aws-codedeploy/test/server/deployment-group.test.ts:
```diff
@@ -412,4 +412,90 @@ describe('CodeDeploy Server Deployment Group', () => {
     expect(() => SynthUtils.toCloudFormation(stack)).toThrow(/deploymentInAlarm/);
   });
+
+  test('can be imported with an ALB Target Group as the load balancer', () => {
```
```suggestion
  test('can be used with an imported ALB Target Group as the load balancer', () => {
```

skinny85 commented on packages/@aws-cdk/aws-codedeploy/test/server/deployment-group.test.ts:
```diff
@@ -412,4 +412,90 @@ describe('CodeDeploy Server Deployment Group', () => {
+      ),
+    });
+
+    expect(stack).toHaveResource('AWS::CodeDeploy::DeploymentGroup', {
```
```suggestion
    expect(stack).toHaveResourceLike('AWS::CodeDeploy::DeploymentGroup', {
```

skinny85 commented on packages/@aws-cdk/aws-codedeploy/test/server/deployment-group.test.ts:
```diff
@@ -412,4 +412,90 @@ describe('CodeDeploy Server Deployment Group', () => {
+          },
+        ],
+      },
+      'DeploymentStyle': {
+        'DeploymentOption': 'WITH_TRAFFIC_CONTROL',
+      },
+    });
```
I don't think this is relevant to this test, let's just remove this.

skinny85 commented on packages/@aws-cdk/aws-codedeploy/test/server/deployment-group.test.ts:
```diff
@@ -412,4 +412,90 @@ describe('CodeDeploy Server Deployment Group', () => {
+          },
+        ],
+      },
+      'DeploymentStyle': {
+        'DeploymentOption': 'WITH_TRAFFIC_CONTROL',
+      },
```
I don't think this is relevant to this test, let's just remove this.

skinny85 commented on packages/@aws-cdk/aws-elasticloadbalancingv2/test/alb/target-group.test.ts:
```diff
@@ -497,4 +497,18 @@ describe('tests', () => {
+    });
+
+    // THEN
+    expect(importedTg.targetGroupName).toEqual('myAlbTargetGroup');
```
Is this really correct? Shouldn't the name be `'myAlbTargetGroup/73e2d6bc24d8a067'`?

skinny85 commented on packages/@aws-cdk/aws-elasticloadbalancingv2/test/nlb/target-group.test.ts:
```diff
@@ -499,4 +499,18 @@ describe('tests', () => {
+    });
+
+    // THEN
+    expect(importedTg.targetGroupName).toEqual('myNlbTargetGroup');
```
Is this really correct? Shouldn't the name be `'myAlbTargetGroup/73e2d6bc24d8a067'`?

skinny85 reviewed (changes requested):
Looks good @eastNine, but we need a few changes before we merge this in.

skinny85 commented:
@eastNine make sure to re-request my review (there's a button in the top-right of the PR window, next to my avatar) once you're ready for another round of reviews - this way, I won't miss another iteration when it's ready. Thanks!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: bf4d07e48437a0393c6d653c71ee05231d553551
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=aa32d841-02aa-4a30-a70d-07f79f154893%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

eastNine commented:
@skinny85
Thanks for review!
I updated PR.

```
this.targetGroupName = cdk.Stack.of(scope).splitArn(props.targetGroupArn, cdk.ArnFormat.SLASH_RESOURCE_SLASH_RESOURCE_NAME).resourceName!;
```

I tried above,
it returns ```myAlbTargetGroup/73e2d6bc24d8a067``` as resourcename in test.

I think we can't use return value of ```Stack.splitArn()```.
Because, property targetGroupInfo.Name is only aceept targetgroup's name instead of resourcename.

[Reference aws-properties-codedeploy-deploymentgroup-targetgroupinfo](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codedeploy-deploymentgroup-targetgroupinfo.html)

So I added one more split.

https://github.com/eastNine/aws-cdk/blob/80b9e37d56630b83f70b254920a67c588846f685/packages/%40aws-cdk/aws-elasticloadbalancingv2/lib/shared/imported.ts#L35

eastNine commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-target-group.ts:
```diff
@@ -350,7 +350,12 @@ export interface TargetGroupAttributes {
+  /**
+   * The name of the target group
+   *
+   * @default targetGroupName
```
This code added while my test.
I removed this.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 80b9e37d56630b83f70b254920a67c588846f685
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=134713e0-84d9-47f7-9a37-43dfebf7c777%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: c15e59caa17d2aabc9cd3248fe08f5b23e3bf545
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0d9ca143-0cd4-4ad2-b078-1ed5b4f55dc9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e3823b6d17aff9540fd14e8c4c151ff143a0a937
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc457f0f-1c32-4a08-a961-3eca64017b2b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (approved):
Looks great @eastNine, thanks so much for the contribution!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 13fc16d030942a4bdb948576292d30742224312c
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a5ec3e3d-2833-4ef4-ad74-3e637cd0671a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
